### PR TITLE
New: Rest Parameter (refs: #10)

### PR DESCRIPTION
**Pull request by xjamundx**
- [x] test from harmony ported over
- [x] additional tests
- [x] error tests

### CODE REVIEW

File: tests/lib/ecma-features.js [hunk 1]
```diff
@@ -93,7 +93,11 @@ describe("ecmaFeatures", function() {
 
                 result = ex;    // if an error is thrown, match the error
             }
-            assert.deepEqual(result, expected);
+            try {
+                assert.deepEqual(result, expected);
+            } catch (e) {
+                throw result;
```
Comment (xjamundx) [inline tests/lib/ecma-features.js]: didn't mean to commit this, but it's really helpful for debugging. (will remove :))

File: README.md [hunk 2]
```diff
@@ -89,6 +89,9 @@ var ast = espree.parse(code, {
         // enable parsing of default parameters
         defaultParams: false,
 
+        // enable parsing of rest parameters
+        restParam: true,
```
Comment (nzakas) [inline README.md]: restParams (to match defaultParams)

File: espree.js [hunk 3]
```diff
@@ -517,7 +517,18 @@ function scanPunctuator() {
         }
     }
 
-    // 3-character punctuators: === !== >>> <<= >>=
+    // 3-character punctuators: ... === !== >>> <<= >>=
```
Comment (nzakas) [inline espree.js]: Double check this - pretty sure this code is already somewhere in there because it's used for JSX. Look for ecmaFeatures.jsx

**Comment by xjamundx** [comment]
Just have one broken test now. Some slightly different line numbers for `arrowFunctions`:

```

      AssertionError: expected { Object (type, body, ...) } to deeply equal { Object (type, body, ...) }
      + expected - actual

                       "column": 26
                       "line": 1
                     }
                     "start": {
      +                "column": 19
      -                "column": 20
                       "line": 1
                     }
                   }
                   "range": [
      +              19
      -              20
                     26
                   ]
                   "type": "SequenceExpression"
                 }
```

**Comment by xjamundx** [comment]
Works now and a bunch of `Spread` stuff in there, because there is some overlap. Will attack spread next.

File: lib/ast-node-factory.js [hunk 4]
```diff
@@ -56,19 +56,20 @@ module.exports = {
...
      * @param {ASTNode} defaults Any default arguments
      * @param {ASTNode} body The function body
+     * @param {ASTNode} rest The rest parameter
      * @param {boolean} expression True if the arrow function is created via an expression.
      *      Always false for declarations, but kept here to be in sync with
      *      FunctionExpression objects.
      * @returns {ASTNode} An ASTNode representing the entire arrow function expression
      */
-    createArrowFunctionExpression: function (params, defaults, body, expression) {
+    createArrowFunctionExpression: function (params, defaults, body, rest, expression) {
```
Comment (nzakas) [inline lib/ast-node-factory.js]: What about `FunctionExpression` and `FunctionDeclaration`?

**Comment by nzakas** [comment]
Looks like you're missing changes to support rest params in `FunctionDeclaration` and `FunctionExpression`.

File: lib/ast-node-factory.js (hunk 4 above)
Comment (xjamundx) [inline lib/ast-node-factory.js]: I think it was actually already there, but I'll add some more tests!

**Comment by xjamundx** [comment]
We now have tests for `var x = function(...b) {}` and `function x(...b){}` and a few variations....

**Comment by nzakas** [comment]
Huh, secret implementation details revealed. Looks good!

**Pull request merged**